Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+
+                foreach (IMixedRealitySpatialAwarenessMeshObserver observer in observers)
+                {
+                    return observer.DisplayOption;
+                }
+
+                // if the spatial awareness is not enabled
+                return SpatialAwarenessMeshDisplayOptions.None;   
+            }
+        }
+
+        private IMixedRealityDataProviderAccess spatialAwarenessSystemAcecess;
+
+        [SerializeField]
+        [Tooltip("If false, the game object to place will not change its rotation according to the surface hit.  The object will" +
+            "remain facing the camera while it is in the placing state.  If true, the object will rotate according to the surface normal" +
+            "if there is a hit.")]
+        private bool rotateAccordingToSurface = false;
+
+        /// <summary>
+        /// If false, the game object to place will not change its rotation according to the surface hit.  The object will 
+        /// remain facing the camera while it is in the placing state.  If true, the object will rotate according to the surface normal
+        /// if there is a hit."
+        /// </summary>
+        public bool RotateAccordingToSurface
+        {
+            get => rotateAccordingToSurface;
+            set
+            {
+                if (rotateAccordingToSurface != value)
+                {
+                    rotateAccordingToSurface = value;
+                }
+            }
+        }
+
+        private const int IgnoreRaycastLayer = 2;
+
+        private const int DefaultLayer = 0;
+
+        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint)
+        protected RayStep currentRay;
+
+        protected RaycastHit currentHit;
+
+        protected float previousFrameNumber = 0;
+
+        protected override void Start()
+        {
+            // Solver base class
+            base.Start();
+
+            // If tap to place is added via script, set the GameObjectToPlace as this gameobject 
+            if (GameObjectToPlace == null)
+            {
+                GameObjectToPlace = gameObject;
+            }
+
+            if(!ColliderPresent)
+            {
+                Debug.LogError("The GameObjectToPlace does not have a collider attached, please attach a collider");
+            }
+
+            SurfaceNormalOffset = GetComponent<Collider>().bounds.extents.z;
+
+            // What if the spatial awareness is not enabled?
+            spatialAwarenessSystemAcecess = CoreServices.SpatialAwarenessSystem as IMixedRealityDataProviderAccess;
+
+            // Get the spatial mesh display option on start
+            // I do not know if this is legal, check on this
+            initialSpatialMeshDisplayOption = InitialSpatialMeshDisplayOption;
+
+            if (AutoStart)
+            {
+                StartPlacement();
+            }
+            else
+            {
+                SolverHandler.UpdateSolvers = false;
```

Comment:
If there are other solvers on this object, wouldn't this also disable the other ones? i.e. is it not valid to have this be on a thing (but not auto-start) and only have it start working on demand?

Reply:
Yes this would disable other solvers on an object.  But is there a case where there would be two active solvers on one object? @wiwei

Reply:
Yes there are cases where two solvers can be active on one object. That is actually a semi-fundamental part of the way solvers were originally designed. This class should definitely just disable itself if applicable

Reply:
Yep, they're pretty much designed to be able to be chained (i.e. why there's a solver handler/manager that orders them in a deterministic way)

Ideally this class should just no-op when it's not in placement mode, though FWIW I would imagine that the behavior here is kinda weird when using multiple solvers (i.e. tap to place really does expect to be alone.)